### PR TITLE
add support for reward func using nn.Module in GRPOTrainer

### PR DESCRIPTION
**Pull request by Tavish9**
# What does this PR do?

This PR adds support for `nn.Module` in the reward functions for the GRPOTrainer. The main reasons for adding this support are as follows:

1.The `PreTrainedModel` in the native transforms are themselves `nn.Module`, and the Transformers'Trainer can support an `nn.Module` as the base model for training. Therefore, the reward function can also be an `nn.Module`.

2.In most cases, our reward function does not require a separate processing class, such as those used with `sentence_transformers` or `pytorch` models. We only need to implement a `forward` function that returns a scalar as the reward.

After this, we can now using the easily-defined nn.Module as the reward func, like below:
```python
class CustomModule(nn.Module):
    def __init__(self, model_name):
        super().__init__()

        self.model = SentenceTransformer(model_name, local_files_only=True, device="cpu")

    def forward(self, completions, answer, **kwargs):
        sentences = content_answers + answer
        embeddings = self.model.encode(sentences, convert_to_tensor=True)
        contents_emb, answers_emb = torch.chunk(embeddings, 2)
        sim = self.model.similarity_pairwise(contents_emb, answers_emb)
        return sim.tolist()
```

```python
from trl import GRPOConfig, GRPOTrainer, TrlParser

parser = TrlParser(GRPOConfig)
(training_args,) = parser.parse_args_into_dataclasses()

my_custom_fn = CustomModule("sentence-transformers/all-MiniLM-L6-v2")

trainer = GRPOTrainer(model="Qwen/Qwen2.5-0.5B-Instruct", reward_funcs=[my_custom_fn], args=training_args)
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by qgallouedec** [comment]
Why not using


```python
model = SentenceTransformer(model_name, local_files_only=True, device="cpu")

def reward_func(completions, answer, **kwargs):
    sentences = content_answers + answer
    embeddings = model.encode(sentences, convert_to_tensor=True)
    contents_emb, answers_emb = torch.chunk(embeddings, 2)
    sim = self.model.similarity_pairwise(contents_emb, answers_emb)
    return sim.tolist()

trainer = GRPOTrainer(
    reward_funcs= reward_func,
    ...
```

instead? Is it because you want the model be "prepared" with `self.accelerator.prepare_model` right?

**Comment by Tavish9** [comment]
yes, sometimes the reward model should be put to cuda for faster inference. Besides, init the model outside the reward function seems weird. Wrapping the model into a basic nn.Module matches the python style :)

**Comment by Tavish9** [comment]
To make motivation more clear, 

1. `function` is suitable for rule-based reward, and it’s independent of the model (though model could be used inside function, not recommended)
2. `PreTrainedModel` is designed for commonly used, predefined non-rule-based reward, such as those for math tasks.
3. `nn.Module` is applicable for other scenarios requiring custom models. Additionally, `processing_class` is not needed.

**Comment by Tavish9** [comment]
any suggestions? @qgallouedec

**Comment by Tavish9** [comment]
Update to the latest commit or just abort? @qgallouedec

**Comment by qgallouedec** [comment]
Sorry for the late reply. Please keep this PR open, I'll try to review it again as soon as possible

**Comment by Tavish9** [comment]
okay, then i will keep rebasing to main :)

**Comment by Tavish9** [comment]
It works. But there are two issues,

1. `nn.Modules` shouldn’t force users to use `input_ids` as input. If it does, then what’s the difference between it and a `PreTrainedModel`? I believe `nn.Modules` are meant to conveniently extend `PreTrainedModel`.
2. `trainer.reward_funcs[0] = prepare_deepspeed(trainer.reward_funcs[0], trainer.accelerator)` putting initialization outside trainer seems weird. The device might not aligned with what trainer wants.

**Comment by Tavish9** [comment]
any suggestions? @qgallouedec
